File: agent-cli/packages/pi/src/timing.ts
export function createPhaseTimer(
	enabled: boolean,
	clock: () => number = () => performance.now(),
	write: (line: string) => void = (line) => { process.stderr.write(line); },
): (phase: string) => void {
	if (!enabled) return () => {};
	let previous = 0;
	return (phase) => {
		const elapsed = clock();
		write(`[cloudthinker timing] ${JSON.stringify({ phase, elapsed_ms: elapsed, phase_ms: elapsed - previous })}\n`);
		previous = elapsed;
	};
}

export const markStartup = createPhaseTimer(process.env.CLOUDTHINKER_TIMING === "1");

File: crates/cloudthinker-cli/src/engine/timing.rs
use std::io::Write as _;
use std::time::Instant;

pub struct PhaseTimer {
    start: Option<Instant>,
    previous_ms: f64,
}

impl PhaseTimer {
    pub fn from_env() -> Self {
        Self {
            start: (std::env::var("CLOUDTHINKER_TIMING").as_deref() == Ok("1")).then(Instant::now),
            previous_ms: 0.0,
        }
    }

    pub fn mark(&mut self, phase: &str) {
        if let Some(start) = self.start {
            let elapsed_ms = start.elapsed().as_secs_f64() * 1000.0;
            let _ = writeln!(
                std::io::stderr(),
                "[cloudthinker timing] {{\"phase\":\"{phase}\",\"elapsed_ms\":{elapsed_ms:.3},\"phase_ms\":{:.3}}}",
                elapsed_ms - self.previous_ms
            );
            self.previous_ms = elapsed_ms;
        }
    }
}
